packages/entity-full-integration-tests/src/__integration-tests__/EntityCacheInconsistency-test.ts

```diff
@@ -0,0 +1,190 @@
+
+  await knex.schema.createTable('testentities', (table) => {
+    table.uuid('id').defaultTo(knex.raw('uuid_generate_v4()')).primary();
+    table.string('other_string').notNullable();
+    table.string('third_string').notNullable();
+  });
+  await knex.into('testentities').truncate();
+}
+
+async function dropPostgresTable(knex: Knex): Promise<void> {
+  if (await knex.schema.hasTable('children')) {
+    await knex.schema.dropTable('children');
+  }
+  if (await knex.schema.hasTable('parents')) {
+    await knex.schema.dropTable('parents');
+  }
+}
+
+describe('Entity cache inconsistency', () => {
+  let knexInstance: Knex;
+  let redisCacheAdapterContext: RedisCacheAdapterContext;
+
+  beforeAll(() => {
+    knexInstance = Knex({
+      client: 'pg',
+      connection: {
+        user: process.env.PGUSER,
+        password: process.env.PGPASSWORD,
+        host: 'localhost',
+        port: parseInt(process.env.PGPORT!, 10),
+        database: process.env.PGDATABASE,
+      },
+    });
+    redisCacheAdapterContext = {
+      redisClient: new Redis(new URL(process.env.REDIS_URL!).toString()),
+      makeKeyFn(...parts: string[]): string {
+        const delimiter = ':';
+        const escapedParts = parts.map((part) =>
+          part.replace('\\', '\\\\').replace(delimiter, `\\${delimiter}`)
+        );
+        return escapedParts.join(delimiter);
+      },
+      cacheKeyPrefix: 'test-',
+      cacheKeyVersion: 1,
+      ttlSecondsPositive: 86400, // 1 day
+      ttlSecondsNegative: 600, // 10 minutes
+    };
+  });
+
+  beforeEach(async () => {
+    await createOrTruncatePostgresTables(knexInstance);
+    await redisCacheAdapterContext.redisClient.flushdb();
+  });
+
+  afterAll(async () => {
+    await dropPostgresTable(knexInstance);
+    knexInstance.destroy();
+    redisCacheAdapterContext.redisClient.disconnect();
+  });
+
+  test('lots of updates in long-ish running transactions', async () => {
+    const viewerContext = new ViewerContext(
+      createFullIntegrationTestEntityCompanionProvider(knexInstance, redisCacheAdapterContext)
+    );
+
+    const entity1 = await TestEntity.creator(viewerContext)
+      .setField('other_string', 'hello')
+      .setField('third_string', 'wat')
+      .enforceCreateAsync();
+
+    // put entities in cache and dataloader
+    await TestEntity.loader(viewerContext).enforcing().loadByIDAsync(entity1.getID());
+    await TestEntity.loader(viewerContext)
+      .enforcing()
+      .loadByFieldEqualingAsync('other_string', 'hello');
+
+    await Promise.all([
+      // do a load after the transaction below updates the entity but before transaction commits to ensure
+      // that the cache is cleared after the transaction commits rather than in the middle where the changes
+      // may not be visible by other requests (ViewerContexts) which would cache the incorrect
+      // value during the read-through cache.
+      (async () => {
+        await new Promise((r) => setTimeout(r, 100));
```

Comment:
For controlling the exact order of events I find explicit barriers to be useful. Something simple like:

```ts
let openBarrier1;
const barrier1 = new Promise((resolve) => {
  openBarrier1 = resolve;
});

...

// one task
await barrier1;

...

// the other task
openBarrier1();
```

This way the tests run as fast as possible and the desired order of events is explicitly expressed instead of implicitly with timeout values.